### PR TITLE
[fix]: [PL-24678]: Fix the issue where scm calls are failing intermittently in delegates

### PR DESCRIPTION
**Pull request by DeepakPatankar**
Context: 

The way scm service works on the delegate is

1. We first run a command like `./scm  --unix socketAddress`
2. This creates an SCM service at the socketAddress
3. Then we try to communicate with the scm service.
4. There might be a chance that the service is not up yet and we are trying to communicate with it.
5. So when the service is not up we get a ScmRuntimeException. 
6. We will retry the scm API call if we get this error.

Recently there was a fix, where we started throwing ConnectExpection in place of ScmRuntimeException. Since we are not retrying for ConnectException, we were getting the error that the SCM is not Available. I have added the retry logic for ConnectException also.

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java [hunk 1]
```diff
@@ -55,6 +56,11 @@ public <R> R processScmRequest(Function<Channel, R> functor) {
             shutdownChannel(channel);
             throw e;
           }
+        } catch (ConnectException ex) {
+          if (++retryCount > 20) {
+            throw new InvalidRequestException("Cannot start Scm Unix Manager", ex);
```
Comment (abhinav-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java:61]: This is not an actionable exception for user. we should give some msg which he either understands appended along or something else imo

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: 1 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 61 at r1](https://reviewable.io/reviews/harness/harness-core/32480#-N0Vm8pFaJZz-VkEHfCM-r1-61:-N0ZSNNe8Z3c1vdrh2qv:b-96k9a6) ([raw file](https://github.com/harness/harness-core/blob/ae5325ec00a0ca2af4a318aeb8c3357654b6936a/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L61)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

This is not an actionable exception for user. we should give some msg which he either understands appended along or something else imo
</blockquote></details>


Since this is an Unexpected Exception, I have changed it from Invalid Request -> Unexpected Exception. Also since for unexpected exception, should not show much details to the users which I also fixed

**Review by bhavya181 — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: 1 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by DeepakPatankar** [comment]
trigger messagecheck

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: :shipit: complete! ✓

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_87Sl-TtNqjC5qu8I:-N0_87Sl-TtNqjC5qu8J:b-tlti76) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L53)):*
> ```Java
>           if (e.getStatus().getCode().equals(Status.Code.UNAVAILABLE)) {
>             if (++retryCount > 20) {
>               log.error("Cannot start SCM Unix Manager");
> ```

no logging of exception?

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_89AWAc355CqnamMF:-N0_89AWAc355CqnamMG:b-tlti76) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
> ```Java
>         } catch (ConnectException ex) {
>           if (++retryCount > 20) {
>             log.error("Cannot start SCM Unix Manager");
> ```

no logging of exception?

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 55 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8BUtDN0u7kbGC8SD:-N0_8BUtDN0u7kbGC8SE:b-qh8rde) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L55)):*
> ```Java
>               log.error("Cannot start SCM Unix Manager");
>               throw new UnexpectedException(
>                   "We are not able to communicate with a internal service. Please contact Harness Support Team", e);
> ```

We should just mention something like "Faced internal server error, please contanct....".

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8vJJ8_Kj_BwwBQ5g:-N0_8vJJ8_Kj_BwwBQ5h:b-smok4i) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L53)):*
> ```Java
>           if (e.getStatus().getCode().equals(Status.Code.UNAVAILABLE)) {
>             if (++retryCount > 20) {
>               log.error("Cannot start SCM Unix Manager");
> ```

This log should be there for each failure, right?

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8yrbE1L0KBpo96Nd:-N0_8yrbE1L0KBpo96Ne:b-smok4i) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
> ```Java
>         } catch (ConnectException ex) {
>           if (++retryCount > 20) {
>             log.error("Cannot start SCM Unix Manager");
> ```

This log should be there for each failure, right?

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: 5. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_87Sl-TtNqjC5qu8I:-N0_C4hH5zVovyFus8gY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L53)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

no logging of exception?
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8vJJ8_Kj_BwwBQ5g:-N0_C5oQ6yEwR_REckBv:bsyg9jc) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L53)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This log should be there for each failure, right?
</blockquote></details>

Added log for each exception
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 55 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8BUtDN0u7kbGC8SD:-N0_IaVuAy56Lw5Os5Ht:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L55)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

We should just mention something like "Faced internal server error, please contanct....". 
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_89AWAc355CqnamMF:-N0_IfC26xIKLnVXzQIv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

no logging of exception?
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8yrbE1L0KBpo96Nd:-N0_IgHJ4eX6tOW4AuMm:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This log should be there for each failure, right?
</blockquote></details>

Done.

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 51 at r3](https://reviewable.io/reviews/harness/harness-core/32480#-N0_LUd77JZMY0sPRClA:-N0_LUd77JZMY0sPRClB:bshtiur) ([raw file](https://github.com/harness/harness-core/blob/c5e84d9f354000b20b800b728335a257fe38d6ce/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L51)):*
> ```Java
>           return functor.apply(channel);
>         } catch (StatusRuntimeException e) {
>           log.error("Error while communicating with the scm service", e);
> ```

can we add retry count as well into it, that will give us more info.

**Comment by mohitg0795** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8yrbE1L0KBpo96Nd:-N0_MF0q52WGmudxEFPF:bshtiur) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
<details><summary><i>Previously, DeepakPatankar (Deepak Patankar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

can we add retry count as well into it, that will give us more info.

**Review by DeepakPatankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: 1 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/32480#-N0_8yrbE1L0KBpo96Nd:-N0_Mrvw6pzGb6r6AXfo:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/66d3b8ddc0c63f395229399b60a48ef2c6b1ab48/930-delegate-tasks/src/main/java/io/harness/delegate/task/scm/ScmDelegateClientImpl.java#L64)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

can we add retry count as well into it, that will give us more info.
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480#-:-N0_OMGY9vpYCJvHYHd_:bnfp4nl)* status: 1 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480#-:-N0_ORQ8AXPYAjhaW6Gb:bnfp4nl)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by bhavya181 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32480)* status: :shipit: complete! ✓

**Pull request merged**